### DISCUSSION TITLE
Issue caused by functionCache in the Function class

### DISCUSSION THREAD

**Discussion by chsword** [RageAgainstThePixel/OpenAI-DotNet#289] [Q&A] [answered]
# Bug Report

Issue caused by functionCache in the Function class: 

In the same project, when using the same Function.Name in different processes with different parameters and descriptions, the functionCache causes this situation to be unhandled.

**Comment by StephenHodgson** [comment, marked as answer]
You cannot use the same name for different functions. They should have unique names

**Reply by chsword** [reply]
If The definition of the function is loaded from the database. If we use static to cache the function definition, we would need to clear the cache every time the function is modified.

**Reply by StephenHodgson** [reply]
@chsword you can use `Tool.ClearRegisteredTools();`

But tbh if you're using a database you probably ought to also be using cloud functions

**Comment by Muleskinner** [comment]
I have the same issue which started to occur after the latest update. This for example now throws the error "The function `functionname` is already registered" when called twice from two different processes:

```csharp
List<Function> f = new List<Function>
    {
    new Function(
        functionName,
        functionDescription,
        parameters
    )
 };
```
I noticed that the GetOrCreateFunction method, which could help me avoid this issue, is defined as internal. Is there a recommended way to check if a function with a given name already exists in the function cache, considering the method is internal?

**Reply by StephenHodgson** [reply]
I think it makes sense to possibly make that `GetOrCreateFunction` public

**Reply by StephenHodgson** [reply]
@Muleskinner I will not be making `Function.GetOrCreateFunction` public, but you can use `Tool.GetOrCreateTool` which is more appropriate.

**Reply by StephenHodgson** [reply]
Also you should be using a list of Tools not Functions:

```csharp
List<Tool> tools = new List<Tools>
{
    new Function(
        functionName,
        functionDescription,
        parameters)
};
 ```
